File: backend/InventoryApi/Services/SupplierService.cs
using InventoryApi.Data;
using InventoryApi.Models;
using InventoryApi.Services.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace InventoryApi.Services;

public class SupplierService : ISupplierService
{
    private readonly AppDbContext _db;
    public SupplierService(AppDbContext db) => _db = db;

    public async Task<IEnumerable<Supplier>> GetAll() => await _db.Suppliers.OrderBy(s => s.Name).ToListAsync();

    public async Task<Supplier?> Get(int id) => await _db.Suppliers.FindAsync(id);

    public async Task<Supplier> Create(string name, string? contactName, string? phone, string? email, string? address)
    {
        var supplier = new Supplier
        {
            Name = name,
            ContactName = contactName,
            Phone = phone,
            Email = email,
            Address = address
        };
        _db.Suppliers.Add(supplier);
        await _db.SaveChangesAsync();
        return supplier;
    }

    public async Task<bool> Update(int id, string name, string? contactName, string? phone, string? email, string? address)
    {
        var supplier = await _db.Suppliers.FindAsync(id);
        if (supplier is null) return false;
        supplier.Name = name;
        supplier.ContactName = contactName;
        supplier.Phone = phone;
        supplier.Email = email;
        supplier.Address = address;
        await _db.SaveChangesAsync();
        return true;
    }

    public async Task<bool> Delete(int id)
    {
        var supplier = await _db.Suppliers.FindAsync(id);
        if (supplier is null) return false;
        _db.Suppliers.Remove(supplier);
        await _db.SaveChangesAsync();
        return true;
    }
}

File: frontend/inventory-app/src/app/services/supplier.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Supplier } from '../models/models';
import { ConfigService } from './config.service';

export interface SupplierDto {
  name: string;
  contactName?: string | null;
  phone?: string | null;
  email?: string | null;
  address?: string | null;
}

@Injectable({ providedIn: 'root' })
export class SupplierService {
  private get baseUrl(): string {
    return `${this.config.apiBaseUrl.replace(/\/$/, '')}/suppliers`;
  }

  constructor(private http: HttpClient, private config: ConfigService) {}

  getAll(): Observable<Supplier[]> {
    return this.http.get<Supplier[]>(this.baseUrl);
  }

  get(id: number): Observable<Supplier> {
    return this.http.get<Supplier>(`${this.baseUrl}/${id}`);
  }

  create(payload: SupplierDto): Observable<Supplier> {
    return this.http.post<Supplier>(this.baseUrl, payload);
  }

  update(id: number, payload: SupplierDto): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/${id}`, payload);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}
